README.md
# fletcher

![CI](https://github.com/xhochy/fletcher/workflows/CI/badge.svg)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/xhochy/fletcher/master)

A library that provides a generic set of Pandas ExtensionDType/Array
implementations backed by Apache Arrow. They support a wider range of types
than Pandas natively supports and also bring a different set of constraints and
behaviours that are beneficial in many situations.

## Usage

To use `fletcher` in Pandas DataFrames, all you need to do is to wrap your data
in a `FletcherChunkedArray` or `FletcherContinuousArray` object. Your data can 
be of either `pyarrow.Array`, `pyarrow.ChunkedArray` or a type that can be passed
to `pyarrow.array(…)`.


```
import fletcher as fr
import pandas as pd

df = pd.DataFrame({
    'str_chunked': fr.FletcherChunkedArray(['a', 'b', 'c']),
    'str_continuous': fr.FletcherContinuousArray(['a', 'b', 'c']),
})

df.info()

# <class 'pandas.core.frame.DataFrame'>
# RangeIndex: 3 entries, 0 to 2
# Data columns (total 2 columns):
#  #   Column          Non-Null Count  Dtype                      
# ---  ------          --------------  -----                      
#  0   str_chunked     3 non-null      fletcher_chunked[string]   
#  1   str_continuous  3 non-null      fletcher_continuous[string]
# dtypes: fletcher_chunked[string](1), fletcher_continuous[string](1)
# memory usage: 166.0 bytes
```

## Development

While you can use `fletcher` in pip-based environments, we strongly recommend
using a `conda` based development setup with packages from `conda-forge`.

```
# Create the conda environment with all necessary dependencies
conda env create

# Activate the newly created environment
conda activate fletcher

# Install fletcher into the current environment
python -m pip install -e . --no-build-isolation --no-use-pep517

# Run the unit tests (you should do this several times during development)
py.test -nauto

# Install pre-commit hooks
# These will then be automatically run on every commit and ensure that files
# are black formatted, have no flake8 issues and mypy checks the type consistency.
pre-commit install
```

Code formatting is done using black. This should keep everything in a
consistent styling and the formatting is automatically adjusted via the
pre-commit hooks.

### Using pandas in development mode

To test and develop against pandas' master or your local fixes, you can install a development version of pandas using:

```
git clone https://github.com/pandas-dev/pandas
cd pandas

# Install additional pandas dependencies
conda install -y cython

# Build and install pandas
python setup.py build_ext --inplace -j 4
python -m pip install -e . --no-build-isolation --no-use-pep517
```

This links the development version of `pandas` into your `fletcher` conda environment.
If you change any Python code in pandas, it is directly reflected in your environment.
If you change any Cython code in pandas, you need to re-execute `python setup.py build_ext --inplace -j 4`.

### Using (py)arrow nightlies

To test and develop against the latest development version of Apache Arrow (`pyarrow`), you can install it from the `arrow-nightlies` conda channel:

```
conda install -c arrow-nightlies arrow-cpp pyarrow
```

### Benchmarks

In `benchmarks/` we provide a set of benchmarks to compare the performance of
`fletcher` against `pandas` and ensure that `fletcher` itself stays performant.
The benchmarks are written using
[airspeed velocity](https://asv.readthedocs.io/en/stable/). When developing
the benchmarks you can run them using `asv dev` (use `-b <pattern>` to only
run a selection of them) only once. To get real benchmark values, you should
use `asv run --python=same` to run the benchmarks multiple times and get
meaningful average runtimes.


CHANGELOG.md
Changelog
=========

Starting with 0.5, we will follow the following versioning scheme:

* We don't bump MAJOR yet.
* We bump MINOR on breaking changes.
* We increase PATCH otherwise.

0.7.2 (2021-01-17)
------------------

* Allow NumPy ufuncs to work with `np.ndarray` outputs where operations are clearly defined (i.e. the fletcher array has no nulls).

0.7.1 (2020-12-29)
------------------

* Fix return values for `str` functions with `pandas=1.2` and `pyarrow=1`.
* Ensure that parallel variants of `apply_binary_str` actually parallize.

0.7.0 (2020-12-07)
------------------

* Add tests for all `str` functions.
* Fix tests for `pyarrow=0.17.1` and add CI jobs for `0.17.1` and `1.0.1`.
* Implement a faster take for list arrays.
* Use `utf8_is_*` functions from Apache Arrow if available.
* Simplify `factorize` implementation to work for chunked arrays with more or less than a single chunk.
* Switch to `pandas.NA` as the user-facing null value
* Add convenience function `fletcher.algorithms.string.apply_binary_str` to apply a binary function on two string columns.

0.6.2 (2020-10-20)
------------------

* Return correct index in functions like `fr_str.extractall`.

0.6.1 (2020-10-14)
------------------

* Create a shallow copy on `.astype(equal dtype, copy=True)`.
* Import `pad_1d` only in older `pandas` versions, otherwise use `get_fill_func`
* Handle `fr_str.extractall` and similar functions correctly, returning a `pd.Dataframe` containing accoring `fletcher` array types.

0.6.0 (2020-09-23)
------------------

* Use `binary_contains_exact` if available from `pyarrow` instead of our own numba-based implementation.
* Provide two more consistent accessors:
 * `.fr_strx`: Call efficient string functions on `fletcher` arrays, error if not available.
 * `.fr_str`: Call string functions on `fletcher` and `object`-typed arrays, convert to `object` if no `fletcher` function is available.
* Add a numba-based implementation for `strip`, `slice`, and `replace`.
* Support `LargeListArray` as a backing structure for lists.
* Implement `isnan` ufunc.

0.5.1 (2020-09-21)
------------------

* Release the GIL where possible.
* Register with dask's `make_array_nonempty` to be able to handle the extension types in `dask`.

0.5.0 (2020-06-23)
------------------

* Implement `FletcherBaseArray.__or__` and `FletcherBaseArray.__any__` to support `pandas.Series.replace`.

0.4.0 (2020-06-16)
------------------

* Forward the `__array__` protocol directly to Arrow
* Add naive implementation for `zfill`
* Add efficient (Numba-based) implementations for `endswith`, `startswith` and `contains`

0.3.1
-----

* Support roundtrips of `pandas.DataFrame` instances with `fletcher` columns through `pyarrow` data structures.
* Move CI to Github Actions

0.3.0
-----

Major changes:
 * We now provide two different extension array implementations.
   There now is the more simpler `FletcherContinuousArray` which is backed by a `pyarrow.Array` instance and thus is always a continuous memory segments.
   The initial `FletcherArray` which is backed by a `pyarrow.ChunkedArray` is now renamed to `FletcherChunkedArray`.
   While `pyarrow.ChunkedArray` allows for more flexibility on how the data is stored, the implementation of algorithms is more complex for it.
   As this hinders contributions and also the adoption in downstream libraries, we now provide both implementations with an equal level of support.
   We don't provide the more general named class `FletcherArray` anymore as there is not a clear opinion on whether this should point to `FletcherContinuousArray` or `FletcherChunkedArray`.
   As usage increases, we might provide such an alias class in future again.
 * Support for `ArithmeticOps` and `ComparisonOps` on numerical data as well as numeric reductions such as `sum`.
   This should allow the use of nullable int and float type for many use cases.
   Performance of nullable integeter columns is on the same level as in `pandas.IntegerArray` as we have similar implementations of the masked arithmetic.
   In future versions, we plan to delegate the workload into the C++ code of `pyarrow` and expect significant performance improvements though the usage of bitmasks over bytemasks.
 * `any` and `all` are now efficiently implemented on boolean arrays.
   We [blogged about this](https://uwekorn.com/2019/09/02/boolean-array-with-missings.html) and how its performance is about twice as fast while only using 1/16 - 1/32 of RAM as the reference boolean array with missing in `pandas`.
   This is due to the fact that prior to `pandas=1.0` you have had to use a float array to have a boolean array that can deal with missing values.
   In `pandas=1.0` a new `BooleanArray` class was added that improves this stituation but also change a bit of the logic.
   We will adapt to this class in the next release and also publish new benchmarks.

New features / performance improvements:
 * For `FletcherContinuousArray` in general and all `FletcherChunkedArray` instances with a single chunk, we now provide an efficient implementation of `take`.
 * Support for Python 3.8 and Pandas 1.0
 * We now check typing in CI using `mypy` and have annotated the code with type hints.
   We only plan to mark the packages as `py.typed` when `pandas` is also marked as `py.typed`.
 * You can query `fletcher` for its version via `fletcher.__version__`
 * Implemented `.str.cat` as `.fr_strx.cat` for arrays with `pa.string()` dtype.
 * `unique` is now supported on all array types where `pyarrow` provides a `unique` implementation.

0.2.0
-----

 * Drop Python 2 support
 * Support for Python 3.7
 * Fixed handling of `date` columns due to new default behaviours in `pyarrow`.

0.1.2
-----

Rerelease with the sole purpose of rendering MarkDown on PyPI.

0.1.1
-----

Load the README in setup.py to have a description on PyPI.

0.1.0
-----

Initial release of fletcher that is based on Pandas 0.23.3 and Apache Arrow 0.9.

benchmarks/text.py
from typing import Any

import numpy as np
import pandas as pd
import pyarrow as pa
import six

import fletcher as fr


def generate_test_array(n):
    return [
        six.text_type(x) + six.text_type(x) + six.text_type(x) if x % 7 == 0 else None
        for x in range(n)
    ]


class TimeSuitePatterns:
    """
    Special benchmark suite for string algorithms looking for patterns.
    """

    def setup(self):
        array = [
            ("a" * 50 + "b" if i % 2 == 0 else "c") * 5 + str(i) for i in range(2 ** 16)
        ]
        self.pattern = "a" * 30 + "b"

        self.df = pd.DataFrame({"str": array})
        self.df_ext = pd.DataFrame(
            {"str": fr.FletcherChunkedArray(pa.array(array, pa.string()))}
        )

    def time_count_no_regex(self):
        self.df["str"].str.count(self.pattern)

    def time_count_no_regex_ext(self):
        self.df_ext["str"].str.count(self.pattern)

    def time_count_no_regex_ext_fr(self):
        self.df_ext["str"].fr_str.count(self.pattern, regex=False)

    def time_contains_no_regex(self):
        self.df["str"].str.contains(self.pattern, regex=False)

    def time_contains_no_regex_ext(self):
        self.df_ext["str"].str.contains(self.pattern, regex=False)

    def time_contains_no_regex_ext_fr(self):
        self.df_ext["str"].fr_str.contains(self.pattern, regex=False)

    def time_replace_no_regex(self):
        self.df["str"].str.replace(self.pattern, "bc")

    def time_replace_no_regex_ext(self):
        self.df_ext["str"].str.replace(self.pattern, "bc", regex=False)

    def time_replace_no_regex_ext_fr(self):
        self.df_ext["str"].fr_str.replace(self.pattern, "bc", regex=False)


class TimeSuiteText:
    def setup(self):
        array = generate_test_array(2 ** 17)
        self.df = pd.DataFrame({"str": array})
        self.df_ext = pd.DataFrame(
            {"str": fr.FletcherChunkedArray(pa.array(array, pa.string()))}
        )

    ###########################
    ## General functionality ##
    ###########################

    def time_isnull(self):
        self.df["str"].isnull()

    def time_isnull_ext(self):
        self.df_ext["str"].isnull()

    def time_concat(self):
        pd.concat([self.df["str"]] * 2)

    def time_concat_ext(self):
        pd.concat([self.df_ext["str"]] * 2)

    ###############################
    ## String accessor functions ##
    ###############################

    def time_capitalize(self):
        self.df["str"].str.capitalize()

    def time_capitalize_ext(self):
        self.df_ext["str"].str.capitalize()

    def time_capitalize_ext_fr(self):
        self.df_ext["str"].fr_str.capitalize()

    def time_casefold(self):
        self.df["str"].str.casefold()

    def time_casefold_ext(self):
        self.df_ext["str"].str.casefold()

    def time_casefold_ext_fr(self):
        self.df_ext["str"].fr_str.casefold()

    def time_cat(self):
        self.df["str"].str.cat(self.df["str"])

    def time_cat_ext(self):
        self.df_ext["str"].str.cat(self.df["str"])

    def time_cat_ext_fr(self):
        self.df_ext["str"].fr_str.cat(self.df_ext["str"])

    def time_contains_no_regex(self):
        self.df["str"].str.contains("1012102", regex=False)

    def time_contains_no_regex_ext(self):
        self.df_ext["str"].str.contains("1012102", regex=False)

    def time_contains_no_regex_ext_fr(self):
        self.df_ext["str"].fr_str.contains("1012102", regex=False)

    def time_contains_no_regex_ignore_case(self):
        self.df["str"].str.contains("0", regex=False, case=False)

    def time_contains_no_regex_ignore_case_ext(self):
        self.df_ext["str"].str.contains("0", regex=False, case=False)

    def time_contains_no_regex_ignore_case_ext_fr(self):
        self.df_ext["str"].fr_str.contains("0", regex=False, case=False)

    def time_contains_regex(self):
        self.df["str"].str.contains("[0-3]", regex=True)

    def time_contains_regex_ext(self):
        self.df_ext["str"].str.contains("[0-3]", regex=True)

    def time_contains_regex_ext_fr(self):
        self.df_ext["str"].fr_str.contains("[0-3]", regex=True)

    def time_contains_regex_ignore_case(self):
        self.df["str"].str.contains("[0-3]", regex=True, case=False)

    def time_contains_regex_ignore_case_ext(self):
        self.df_ext["str"].str.contains("[0-3]", regex=True, case=False)

    def time_contains_regex_ignore_case_ext_fr(self):
        self.df_ext["str"].fr_str.contains("[0-3]", regex=True, case=False)

    def time_count_no_regex(self):
        self.df["str"].str.count("001")

    def time_count_no_regex_ext(self):
        self.df_ext["str"].str.count("001")

    def time_count_no_regex_ext_fr(self):
        self.df_ext["str"].fr_str.count("001", regex=False)

    def time_count_regex(self):
        self.df["str"].str.count("[0-3]")

    def time_count_regex_ext(self):
        self.df_ext["str"].str.count("[0-3]")

    def time_count_regex_ext_fr(self):
        self.df_ext["str"].fr_str.count("[0-3]", regex=True)

    def time_endswith(self):
        self.df["str"].str.endswith("10")

    def time_endswith_ext(self):
        self.df_ext["str"].str.endswith("10")

    def time_endswith_ext_fr(self):
        self.df_ext["str"].fr_str.endswith("10")

    def time_extract(self):
        self.df["str"].str.extract("([0-3]+)")

    def time_extract_ext(self):
        self.df_ext["str"].str.extract("([0-3]+)")

    def time_extract_ext_fr(self):
        self.df_ext["str"].fr_str.extract("([0-3]+)")

    def time_extractall(self):
        self.df["str"].str.extractall("([0-3]+)")

    def time_extractall_ext(self):
        self.df_ext["str"].str.extractall("([0-3]+)")

    def time_extractall_ext_fr(self):
        self.df_ext["str"].fr_str.extractall("([0-3]+)")

    def time_find(self):
        self.df["str"].str.find("10")

    def time_find_ext(self):
        self.df_ext["str"].str.find("10")

    def time_find_ext_fr(self):
        self.df_ext["str"].fr_str.find("10")

    def time_findall(self):
        self.df["str"].str.findall("([0-3]+)")

    def time_findall_ext(self):
        self.df_ext["str"].str.findall("([0-3]+)")

    def time_findall_ext_fr(self):
        self.df_ext["str"].fr_str.findall("([0-3]+)")

    def time_get(self):
        self.df["str"].str.get(2)

    def time_get_ext(self):
        self.df_ext["str"].str.get(2)

    def time_get_ext_fr(self):
        self.df_ext["str"].fr_str.get(2)

    # We don't benchmark index as it is the same as find but just raising errors instead of returning -1

    def time_len(self):
        self.df["str"].str.len()

    def time_len_ext(self):
        self.df_ext["str"].str.len()

    def time_len_ext_fr(self):
        self.df_ext["str"].fr_str.len()

    def time_ljust(self):
        self.df["str"].str.ljust(10)

    def time_ljust_ext(self):
        self.df_ext["str"].str.ljust(10)

    def time_ljust_ext_fr(self):
        self.df_ext["str"].fr_str.ljust(10)

    def time_lower(self):
        self.df["str"].str.lower()

    def time_lower_ext(self):
        self.df_ext["str"].str.lower()

    def time_lower_ext_fr(self):
        self.df_ext["str"].fr_str.lower()

    def time_lstrip(self):
        self.df["str"].str.lstrip("0")

    def time_lstrip_ext(self):
        self.df_ext["str"].str.lstrip("0")

    def time_lstrip_ext_fr(self):
        self.df_ext["str"].fr_str.lstrip("0")

    def time_match(self):
        self.df["str"].str.match("([0-3]+)")

    def time_match_ext(self):
        self.df_ext["str"].str.match("([0-3]+)")

    def time_match_ext_fr(self):
        self.df_ext["str"].fr_str.match("([0-3]+)")

    def time_normalize(self):
        self.df["str"].str.normalize(form="NFC")

    def time_normalize_ext(self):
        self.df_ext["str"].str.normalize(form="NFC")

    def time_normalize_ext_fr(self):
        self.df_ext["str"].fr_str.normalize(form="NFC")

    # We skip pad as it is just center/rjust/ljust

    def time_partition(self):
        self.df["str"].str.partition("0")

    # Not supported yet, see the corresponding unit test
    # def time_partition_ext(self):
    #     self.df_ext["str"].str.partition("0")

    # Not supported yet, see the corresponding unit test
    # def time_partition_ext_fr(self):
    #    self.df_ext["str"].fr_str.partition("0")

    def time_repeat(self):
        self.df["str"].str.repeat(3)

    def time_repeat_ext(self):
        self.df_ext["str"].str.repeat(3)

    def time_repeat_ext_fr(self):
        self.df_ext["str"].fr_str.repeat(3)

    def time_replace_no_regex(self):
        self.df["str"].str.replace("001", "23", regex=False)

    def time_replace_no_regex_ext(self):
        self.df_ext["str"].str.replace("001", "23", regex=False)

    def time_replace_no_regex_ext_fr(self):
        self.df_ext["str"].fr_str.replace("001", "23", regex=False)

    def time_replace_regex(self):
        self.df["str"].str.replace("[0-3]", "5", regex=True)

    def time_replace_regex_ext(self):
        self.df_ext["str"].str.replace("[0-3]", "5", regex=True)

    def time_replace_regex_ext_fr(self):
        self.df_ext["str"].fr_str.replace("[0-3]", "5", regex=False)

    def time_rfind(self):
        self.df["str"].str.rfind("10")

    def time_rfind_ext(self):
        self.df_ext["str"].str.rfind("10")

    def time_rfind_ext_fr(self):
        self.df_ext["str"].fr_str.rfind("10")

    def time_rjust(self):
        self.df["str"].str.rjust(10)

    def time_rjust_ext(self):
        self.df_ext["str"].str.rjust(10)

    def time_rjust_ext_fr(self):
        self.df_ext["str"].fr_str.rjust(10)

    def time_rpartition(self):
        self.df["str"].str.rpartition("0")

    # Not supported yet, see the corresponding unit test
    # def time_rpartition_ext(self):
    #     self.df_ext["str"].str.rpartition("0")

    # Not supported yet, see the corresponding unit test
    # def time_rpartition_ext_fr(self):
    #    self.df_ext["str"].fr_str.rpartition("0")

    def time_rstrip(self):
        self.df["str"].str.rstrip("0")

    def time_rstrip_ext(self):
        self.df_ext["str"].str.rstrip("0")

    def time_rstrip_ext_fr(self):
        self.df_ext["str"].fr_str.rstrip("0")

    def time_slice(self):
        self.df["str"].str.slice(2, 5)

    def time_slice_ext(self):
        self.df_ext["str"].str.slice(2, 5)

    # FIXME: Timeouts
    # def time_slice_ext_fr(self):
    #     self.df_ext["str"].fr_str.slice(2, 5)

    def time_slice_step2(self):
        self.df["str"].str.slice(2, 5, 2)

    def time_slice_step2_ext(self):
        self.df_ext["str"].str.slice(2, 5, 2)

    # FIXME: Timeouts
    # def time_slice_step2_ext_fr(self):
    #     self.df_ext["str"].fr_str.slice(2, 5, 2)

    def time_slice_replace(self):
        self.df["str"].str.slice_replace(start=1, stop=3, repl="X")

    def time_slice_replace_ext(self):
        self.df_ext["str"].str.slice_replace(start=1, stop=3, repl="X")

    def time_slice_replace_ext_fr(self):
        self.df_ext["str"].fr_str.slice_replace(start=1, stop=3, repl="X")

    def time_split(self):
        self.df["str"].str.split("0")

    def time_split_ext(self):
        self.df_ext["str"].str.split("0")

    def time_split_ext_fr(self):
        self.df_ext["str"].fr_str.split("0")

    def time_rsplit(self):
        self.df["str"].str.rsplit("0")

    def time_rsplit_ext(self):
        self.df_ext["str"].str.split("0")

    def time_rsplit_ext_fr(self):
        self.df_ext["str"].fr_str.rsplit("0")

    def time_startswith(self):
        self.df["str"].str.startswith("10")

    def time_startswith_ext(self):
        self.df_ext["str"].str.startswith("10")

    def time_startswith_ext_fr(self):
        self.df_ext["str"].fr_str.startswith("10")

    def time_strip(self):
        self.df["str"].str.strip("0")

    def time_strip_ext(self):
        self.df_ext["str"].str.strip("0")

    # FIXME: timeouts
    # def time_strip_ext_fr(self):
    #     self.df_ext["str"].fr_str.strip('0')

    def time_swapcase(self):
        self.df["str"].str.swapcase()

    def time_swapcase_ext(self):
        self.df_ext["str"].str.swapcase()

    def time_swapcase_ext_fr(self):
        self.df_ext["str"].fr_str.swapcase()

    def time_title(self):
        self.df["str"].str.title()

    def time_title_ext(self):
        self.df_ext["str"].str.title()

    def time_title_ext_fr(self):
        self.df_ext["str"].fr_str.title()

    def time_translate(self):
        self.df["str"].str.translate({"0": "9", "1": "8", "2": "7", "3": "6", "4": "5"})

    def time_translate_ext(self):
        self.df_ext["str"].str.translate(
            {"0": "9", "1": "8", "2": "7", "3": "6", "4": "5"}
        )

    def time_translate_ext_fr(self):
        self.df_ext["str"].fr_str.translate(
            {"0": "9", "1": "8", "2": "7", "3": "6", "4": "5"}
        )

    def time_upper(self):
        self.df["str"].str.upper()

    def time_upper_ext(self):
        self.df_ext["str"].str.upper()

    def time_upper_ext_fr(self):
        self.df_ext["str"].fr_str.upper()

    def time_wrap(self):
        self.df["str"].str.wrap(5)

    def time_wrap_ext(self):
        self.df_ext["str"].str.wrap(5)

    def time_wrap_ext_fr(self):
        self.df_ext["str"].fr_str.wrap(5)

    def time_zfill(self):
        self.df["str"].str.zfill(10)

    def time_zfill_ext(self):
        self.df_ext["str"].str.zfill(10)

    def time_zfill_ext_fr(self):
        self.df_ext["str"].fr_str.zfill(10)

    def time_isalnum(self):
        self.df["str"].str.isalnum()

    def time_isalnum_ext(self):
        self.df_ext["str"].str.isalnum()

    def time_isalnum_ext_fr(self):
        self.df_ext["str"].fr_str.isalnum()

    def time_isalpha(self):
        self.df["str"].str.isalpha()

    def time_isalpha_ext(self):
        self.df_ext["str"].str.isalpha()

    def time_isalpha_ext_fr(self):
        self.df_ext["str"].fr_str.isalpha()

    def time_isdigit(self):
        self.df["str"].str.isdigit()

    def time_isdigit_ext(self):
        self.df_ext["str"].str.isdigit()

    def time_isdigit_ext_fr(self):
        self.df_ext["str"].fr_str.isdigit()

    def time_isspace(self):
        self.df["str"].str.isspace()

    def time_isspace_ext(self):
        self.df_ext["str"].str.isspace()

    def time_isspace_ext_fr(self):
        self.df_ext["str"].fr_str.isspace()

    def time_islower(self):
        self.df["str"].str.islower()

    def time_islower_ext(self):
        self.df_ext["str"].str.islower()

    def time_islower_ext_fr(self):
        self.df_ext["str"].fr_str.islower()

    def time_isupper(self):
        self.df["str"].str.isupper()

    def time_isupper_ext(self):
        self.df_ext["str"].str.isupper()

    def time_isupper_ext_fr(self):
        self.df_ext["str"].fr_str.isupper()

    def time_istitle(self):
        self.df["str"].str.istitle()

    def time_istitle_ext(self):
        self.df_ext["str"].str.istitle()

    def time_istitle_ext_fr(self):
        self.df_ext["str"].fr_str.istitle()

    def time_isnumeric(self):
        self.df["str"].str.isnumeric()

    def time_isnumeric_ext(self):
        self.df_ext["str"].str.isnumeric()

    def time_isnumeric_ext_fr(self):
        self.df_ext["str"].fr_str.isnumeric()

    def time_isdecimal(self):
        self.df["str"].str.isdecimal()

    def time_isdecimal_ext(self):
        self.df_ext["str"].str.isdecimal()

    def time_isdecimal_ext_fr(self):
        self.df_ext["str"].fr_str.isdecimal()


class Dummies:
    def setup(self):
        self.s = pd.Series(pd._testing.makeStringIndex(10 ** 5)).str.join("|")
        self.s_ext = pd.Series(
            fr.FletcherChunkedArray(pa.array(self.s.to_numpy(), pa.string()))
        )

    def time_get_dummies(self):
        self.s.str.get_dummies("|")

    def time_get_dummies_ext(self):
        self.s_ext.str.get_dummies("|")

    def time_get_dummies_ext_fr(self):
        self.s_ext.fr_str.get_dummies("|")


class Indexing:
    # index and value have diverse values, disable type checks for them
    indexer: Any
    value: Any

    n = 2 ** 12

    params = [
        (True, False),
        ("scalar_value", "array_value"),
        ("int", "int_array", "bool_array", "slice"),
    ]
    param_names = ["chunked", "values", "indices"]

    def setup(self, chunked, value, indices):
        # assert np.isscalar(values) or len(values) == len(indices)
        array = generate_test_array(self.n)
        if indices == "int":
            if value == "array_value":
                raise NotImplementedError()
            self.indexer = 50
        elif indices == "int_array":
            self.indexer = list(range(0, self.n, 5))
        elif indices == "bool_array":
            self.indexer = np.zeros(self.n, dtype=bool)
            self.indexer[list(range(0, self.n, 5))] = True
        elif indices == "slice":
            self.indexer = slice(0, self.n, 5)

        if value == "scalar_value":
            self.value = "setitem"
        elif value == "array_value":
            self.value = [str(x) for x in range(self.n)]
            self.value = np.array(self.value)[self.indexer]
            if len(self.value) == 1:
                self.value = self.value[0]

        self.df = pd.DataFrame({"str": array})
        if chunked:
            array = np.array_split(array, 1000)
        else:
            array = [array]
        self.df_ext = pd.DataFrame(
            {
                "str": fr.FletcherChunkedArray(
                    pa.chunked_array([pa.array(chunk, pa.string()) for chunk in array])
                )
            }
        )

    def time_getitem(self, chunked, value, indices):
        self.df_ext["str"][self.indexer]

    def time_getitem_obj(self, chunked, value, indices):
        self.df["str"][self.indexer]

    def time_setitem(self, chunked, value, indices):
        self.df_ext["str"][self.indexer] = self.value

    def time_setitem_obj(self, chunked, value, indices):
        self.df["str"][self.indexer] = self.value
